### PR TITLE
Allow migrated domains to redirect form submissions to new host

### PR DESCRIPTION
**Pull request by kaapstorm**
## Product Description

This change allows domains that have been migrated to third-party-managed hosts to redirect form submissions, syncs, case claim requests etc. to their new host.

This avoids problems when deployed CommCare apps are not using a DNS alias prior to the migration.

## Technical Summary

Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field. The redirect URL is set using a management command. While this has the disadvantage of needing to be done by a developer, it has the advantages that a developer can ensure the URL is legitimate, and that the URL can be set after the domain has been locked by the data migration feature flag.

Jira: [SC-3628](https://dimagi.atlassian.net/browse/SC-3628)

Documentation: [Redirect on Data Migration](https://docs.google.com/document/d/1mQxRmboW3CuOkmklUjf5jypgJcv9VX_e26TEiFh1CyI/edit)

FYI @ctsims 

## Feature Flag

DATA_MIGRATION

### Safety story

* HQ behavior is verified with unit tests
* Interaction with mobile to be verified by QA

### Automated test coverage

Tests verify the management command, and the view decorator that protects all URLs for submitting and fetching form and case data.

### QA Plan

QA ticket: [QA-6406](https://dimagi.atlassian.net/browse/QA-6406)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

[SC-3628]: https://dimagi.atlassian.net/browse/SC-3628?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[QA-6406]: https://dimagi.atlassian.net/browse/QA-6406?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Charl1996 — APPROVED** [review]
Looks good to me. 

Just a note: I'm pro streamlining this workflow in the future (not now, since this is a rather high priority ticket), thus having some place in HQ that this could be done, so not having to involve a developer...

File: corehq/apps/domain/decorators.py
Comment (mkangia) [inline corehq/apps/domain/decorators.py:684]: just checking if we want to consider making a redirect and not a permanent redirect?
I don't know if permanent redirect is an absolute requirement.

I say so because if the migration isn't successful and the project wants to start using SaaS environment again, they won't have a way to do that then? Just good to confirm/communicate that

**Review by mkangia — APPROVED** [review]
Great tests.

Just one query that could really be non blocking if it has already been thought about or can't be done.

File: corehq/apps/domain/decorators.py
Comment (kaapstorm) [reply corehq/apps/domain/decorators.py:684]: I agree that we should communicate this very clearly.

They should not use this before the migration is completed, and confirmed to have been successful.

If they enable this before then, then they _will lose data_, and the point of this feature is to avoid them losing data.

File: corehq/apps/domain/models.py
Comment (esoergel) [inline corehq/apps/domain/models.py:453]: > Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field.

No objections here, but I do have some passing idle thoughts.  I have also wondered how best to deal with these random one-off domain settings.  The obvious one would be a `DomainSettings` (sql) model that's got a bunch of columns on it.  There are few enough domains that we don't really need to worry about the scale of data stored.

Another possibility that occurs to me, especially for sparsely populated fields is to use an entity-attribute-value model that doesn't require a migration for new fields.  Something like:
```python
class DomainSetting(models.Model):
    domain = models.CharField(db_index=True)
    slug = models.CharField(db_index=True)
    value = models.HmmActuallyMaybeThisIsHarderThanIThought()
```
Only thing I don't like about that is you'd have to use one fixed schema for `value`, which might be a string or a boolean or something else, depending on the setting.  Okay maybe a simple `DomainSettings` model _is_ best.

Comment (kaapstorm) [reply corehq/apps/domain/models.py]: 675f785d3c4

Comment (mkangia) [reply corehq/apps/domain/models.py]: if we are creating a new model, how about exposing it to django admin?

Comment (kaapstorm) [reply corehq/apps/domain/models.py]: 43efa33b1b9

**Review by mkangia — COMMENTED** [review]
Cool, thanks for the follow up @kaapstorm 

Looks good. Will wait for QA for final approval.

**Comment by mkangia** [comment]
> ^^^ No. The migrations are fast (they just create a table), but they refer to a field type that is defined in the code, and so they need to be applied at the same time.

Curious about this. I believe that check on the PR is to ensure that the migration won't impact the currently running code on the environment. A scenario where till be a concern is if you were to remove a column, that will be referenced on the environment till the code is updated, which would be bad. 
So that check should be okay on this PR since you are just adding a new table which will only be referenced post code update.
Right?

File: corehq/apps/domain/models.py
Comment (millerdev) [inline corehq/apps/domain/models.py:897]: Will this block setting the value to an empty string? Or are validators not executed on empty values?

Also, this seems like it's enforcing a policy decision that may not match real world circumstances. If someone moved their domain to self-hosted and decided not to put an SSL certificate on it, are we not going to redirect to that domain? I think this should be a strong suggestion (warning if violated, probably implemented in UI JS) rather than a hard requirement.

File: migrations.lock
Comment (millerdev) [inline migrations.lock:394]: Can these migrations be squashed? If not, would be nice if the second one had a nicer name.

**Review by millerdev — COMMENTED** [review]
The description says:

> Adds yet another property to the Domain model, but this was deemed better than creating a new model for just one string field.

I'm confused. The current solution in this PR looks like what the description said it was not. Notably, it does not conform with...

Ethan suggested:

> Another possibility that occurs to me ... is to use an entity-attribute-value model that doesn't require a migration for new fields.

I like this suggestion. We could use a `JSONField` to solve the `HmmActuallyMaybeThisIsHarderThanIThought` field type problem. Although that still has a disadvantage of not easily being able to store dates. We could work around that by making an `ExoticValueJsonField` that would auto-convert certain values to rich types (example: `{"__type__": "date", "value": "2024-04-01"}` -> `date(2024, 4, 1)`). Luckily we don't need to solve that problem immediately, so a plain `JSONField` would work for now.

File: corehq/apps/domain/decorators.py
Comment (millerdev) [reply corehq/apps/domain/decorators.py]: But why not make this a non-permanent redirect?

**Comment by millerdev** [comment]
I see one [unanswered question](https://github.com/dimagi/commcare-hq/pull/34461#discussion_r1574788808). Otherwise, I think this is good to go.

File: corehq/apps/domain/decorators.py
Comment (kaapstorm) [reply corehq/apps/domain/decorators.py]: I chose a permanent redirect because once data has been migrated to the new environment, syncing with the old environment will always result in data loss.

We could use a non-permanent redirect (HTTP 302: Found) but I'm not sure that the meaning of the code is one that we want to convey. The device should, from the moment the data migration has completed successfully, ever be syncing with the old environment again. ... Or am I missing a scenario where that is not true?

Comment (kaapstorm) [reply corehq/apps/domain/decorators.py]: Offlined about this. I've switched to a temporary redirect, because people need to be able to change their minds. https://github.com/dimagi/commcare-hq/pull/34461/commits/69488f09934c37513b7ec8727921ff1b4a82c959

**Review by millerdev — APPROVED** [review]

**Review by millerdev — APPROVED** [review]
I'm curious about the things that motivated each of the last three commits. Can you share those details?

**Comment by kaapstorm** [comment]
The mobile team has coalesced on 302.

I've also tweaked the decorator to work when the feature flag is disabled, to allow mobile users to log in, but the feature flag still needs to be enabled to set the redirect URL. So the workflow for data migrations is:
1. Set the Data Migration feature flag
2. Do the actual migration
3. Set the redirect URL to the new environment so that devices can't sync with the old environment.
4. Disable the feature flag so that mobile users can log in.

**Pull request merged**
